### DISCUSSION TITLE
Data table with multi checkboxes

### DISCUSSION THREAD

**Discussion by alnuaimi94** [unpoly/unpoly#373] [Q&A] [answered]
Hi,
I have data table with multi checkboxes:
live demo: [https://spiced-shocking-whistle.glitch.me](https://spiced-shocking-whistle.glitch.me)
code: [https://glitch.com/edit/#!/spiced-shocking-whistle](https://glitch.com/edit/#!/spiced-shocking-whistle)

my problem when i using unpoly and transition between pages the select all checkboxes don't work.

how can solve this problem?

### Updated
i'm using `up-compile` and it's worked finally but i don't no it is correct way?

```javascript
up.compiler("#result_list", () => {
  const defaults = {
    allToggleId: "action-toggle",
    selectedClass: "selected",
  };

  function checker(actionCheckboxes, options, checked) {
    actionCheckboxes.forEach(function (el) {
      el.checked = checked;
      el.closest("tr").classList.toggle(options.selectedClass, checked);
    });
  }

  function updateCounter(actionCheckboxes, options) {
    const sel = Array.from(actionCheckboxes).filter(function (el) {
      return el.checked;
    }).length;
    const allToggle = document.getElementById(options.allToggleId);
    allToggle.checked = sel === actionCheckboxes.length;
  }

  window.Actions = function (actionCheckboxes, options) {
    options = Object.assign({}, defaults, options);
    let list_editable_changed = false;
    let lastChecked = null;
    let shiftPressed = false;

    document.addEventListener("keydown", (event) => {
      shiftPressed = event.shiftKey;
    });

    document.addEventListener("keyup", (event) => {
      shiftPressed = event.shiftKey;
    });

    document
      .getElementById(options.allToggleId)
      .addEventListener("click", function (event) {
        checker(actionCheckboxes, options, this.checked);
        updateCounter(actionCheckboxes, options);
      });

    function affectedCheckboxes(target, withModifier) {
      const multiSelect = lastChecked && withModifier && lastChecked !== target;
      if (!multiSelect) {
        return [target];
      }
      const checkboxes = Array.from(actionCheckboxes);
      const targetIndex = checkboxes.findIndex((el) => el === target);
      const lastCheckedIndex = checkboxes.findIndex((el) => el === lastChecked);
      const startIndex = Math.min(targetIndex, lastCheckedIndex);
      const endIndex = Math.max(targetIndex, lastCheckedIndex);
      const filtered = checkboxes.filter(
        (el, index) => startIndex <= index && index <= endIndex
      );
      return filtered;
    }

    Array.from(document.getElementById("result_list").tBodies).forEach(function (
      el
    ) {
      el.addEventListener("change", function (event) {
        const target = event.target;
        if (target.classList.contains("action-select")) {
          const checkboxes = affectedCheckboxes(target, shiftPressed);
          checker(checkboxes, options, target.checked);
          updateCounter(actionCheckboxes, options);
          lastChecked = target;
        } else {
          list_editable_changed = true;
        }
      });
    });
  };

  const actionsEls = document.querySelectorAll("tr input.action-select");
  if (actionsEls.length > 0) {
    Actions(actionsEls);
  }
});
```

**Comment by adam12** [comment, marked as answer]
You need to use an [`up.compiler`](https://unpoly.com/up.compiler) to detect DOM changes, then setup your event listeners.

This is the change:

```javascript
up.compiler('tr input.action-select', actionEl => {
  Actions([actionEl]);
})
```

Here's my remix.

https://glitch.com/edit/#!/snapdragon-furry-antimony?path=script.js%3A77%3A20

EDIT: I just noticed that you also figured out it was `up.compiler`, since I read the issue a few hours ago. It looks fine but I'd probably try to go more granular; look for the input checkbox and not the entire element.

**Reply by alnuaimi94** [reply]
your method way work fine but when select all checkboxes and try to unselect one of row data the all checkboxes don't updated and remain checked = true; therefore i think we must compile the entire element.

**Reply by adam12** [reply]
Possibly, but I didn't look at the source of your check box code. I'd still try to minimize the scope of the compiler and make it re-usable, which might mean changes to the non-compiler portion of your code.

Take a look at this other remix I just made, that uses Github's check-all library. This might be something you could consider as well if you want. https://glitch.com/edit/#!/mulberry-fragrant-dill?path=about.html%3A27%3A42

**Reply by alnuaimi94** [reply]
Oh, very good
i used this code from django admin action you can see full code here:
[https://github.com/django/django/blob/main/django/contrib/admin/static/admin/js/actions.js](https://github.com/django/django/blob/main/django/contrib/admin/static/admin/js/actions.js)
